packages/flutter_tools/lib/src/android/adb.dart

```diff
@@ -0,0 +1,223 @@
+        logging.fine('--> host:track-devices');
+        socket.add(_createAdbRequest('host:track-devices'));
+        socket.listen((List<int> data) {
+          String stringResult = new String.fromCharCodes(data);
+          logging.fine('<-- ${stringResult.trim()}');
+          _AdbServerResponse response = new _AdbServerResponse(
+            stringResult,
+            noStatus: !isFirstNotification
+          );
+
+          String devicesText = response.message.trim();
+          isFirstNotification = false;
+
+          if (devicesText.isEmpty) {
+            controller.add(<AdbDevice>[]);
+          } else {
+            controller.add(devicesText.split('\n').map((String deviceInfo) {
+              return new AdbDevice(deviceInfo);
+            }).toList());
+          }
+        });
+        socket.done.then((_) => controller.close());
+      },
+      onCancel: () => socket?.destroy()
+    );
+
+    return controller.stream;
+  }
+
+  Future<String> _sendAdbServerCommand(String command) async {
+    Socket socket = await Socket.connect(InternetAddress.LOOPBACK_IP_V4, adbServerPort);
+
+    try {
+      logging.fine('--> $command');
+      socket.add(_createAdbRequest(command));
+      List<List<int>> result = await socket.toList();
+      List<int> data = result.fold(<int>[], (List<int> previous, List<int> element) {
+        return previous..addAll(element);
+      });
+      String stringResult = new String.fromCharCodes(data);
+      logging.fine('<-- ${stringResult.trim()}');
+      return stringResult;
+    } finally {
+      socket.destroy();
+    }
+  }
+}
+
+class AdbDevice {
+  static final RegExp deviceRegex = new RegExp(r'^(\S+)\s+(\S+)(.*)');
+
+  /// Always non-null; something like `TA95000FQA`.
+  String id;
+
+  /// device, offline, unauthorized.
+  String status;
+
+  Map<String, String> _info = <String, String>{};
+
+  AdbDevice(String deviceInfo) {
+    // 'TA95000FQA	device'
+    // 'TA95000FQA             device usb:340787200X product:peregrine_retus model:XT1045 device:peregrine'
+    // '015d172c98400a03       device usb:340787200X product:nakasi model:Nexus_7 device:grouper'
+
+    Match match = deviceRegex.firstMatch(deviceInfo);
+    id = match[1];
+    status = match[2];
+
+    String rest = match[3];
+    if (rest != null && rest.isNotEmpty) {
+      rest = rest.trim();
+      for (String data in rest.split(' ')) {
+        if (data.contains(':')) {
+          List<String> fields = data.split(':');
+          _info[fields[0]] = fields[1];
+        }
+      }
+    }
+  }
+
+  bool get isAvailable => status == 'device';
+
+  /// Device model; can be null. `XT1045`, `Nexus_7`
+  String get modelID => _info['model'];
+
+  /// Device code name; can be null. `peregrine`, `grouper`
+  String get deviceCodeName => _info['device'];
+
+  /// Device product; can be null. `peregrine_retus`, `nakasi`
+  String get productID => _info['product'];
+
+  operator==(other) => other is AdbDevice && other.id == id;
```

Comment:
Please use the same operator== pattern we have elsewhere (explicit return type, space after `operator`, and the specific body we use elsewhere.)

Having just one pattern makes maintaining these much easier.